File: packages/all-in-one/src/rust-utils/Result.ts
export default class Result<T, E extends Error> {
  private readonly success: boolean;

  private readonly data: T | undefined;
  private readonly error: E | undefined;

  protected constructor(success: boolean, data?: T, error?: E) {
    this.success = success;
    this.data = data;
    this.error = error;
  }

  static ok<T, E extends Error>(data: T) {
    return new Result<T, E>(true, data);
  }

  static error<T, E extends Error>(error: E) {
    return new Result<T, E>(false, undefined, error);
  }

  get isOk() {
    return this.success;
  }

  get isError() {
    return !this.success;
  }

  unwrap(): T {
    if (this.success) {
      return this.data as T;
    } else {
      throw this.error;
    }
  }

  catch(): E | null {
    return this.success ? null : (this.error as E); // E 必存在
  }
}

File: crates/test_core/tests/result.rs
use std::io::{Error, ErrorKind};

#[test]
fn smoke_test() {
    let result = Ok::<(), Error>(());
    let result = Err::<(), _>(Error::new(ErrorKind::AlreadyExists, "haha"));
    result.is_ok();
    match result {
        Ok(data) => {}
        Err(err) => {}
    }
}
